src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1493,19 +1493,22 @@
 
 ;; Wrap the fact-type so that Clojure equality and hashcode semantics are used
 ;; even though this is placed in a Java map.
-(deftype AlphaRootsWrapper [roots roots-hash]
+(deftype AlphaRootsWrapper [fact-type fact-type-hash roots]
   Object
   (equals [this other]
     ;; Since we memoize the function that retrieves these on the fact type, and since the alpha roots for
     ;; any given type primary type T (excluding descendants and ancestors) are a disjoint
     ;; set with respect to the alpha roots for any other primary type T2, we know that we will
     ;; only have one AlphaRootsWrapper object for any given set of alpha roots per get-alphas-fn instance.
-    ;; Therefore we can just use reference equality in the interest of maximizing performance.
-    (identical? this other))
+    ;; Therefore we can reference equality in the interest of maximizing performance and fallback to equivalence,
+    ;; as fact-types can be defined by consumers and thus might not be identical.
+    (let [other ^AlphaRootsWrapper other]
+      (and (= fact-type-hash (.fact-type-hash other))
+           (or (identical? fact-type (.fact-type other))
+               (= fact-type (.fact-type other))))))
   ;; Since know we will need to find the hashcode of this object in all cases just eagerly calculate it upfront
-  ;; and avoid extra calls to hash later.  The hashcode of the Clojure vector of roots would be cached but there
-  ;; is no need to even make the call to that cache.
-  (hashCode [this] roots-hash))
+  ;; and avoid extra calls to hash later.
+  (hashCode [this] fact-type-hash))
```

Comment:
In the generic sense of an equals and hashcode contract, since both the equals and hash code depend on the fact type, the equals and hashcode contract would be upheld regardless.  However, in our particular use-case, the AlphaRootsWrapper is being used to logically represent the roots that correspond to a particular concrete fact type that is used in conditions.  There is one-to-one correspondence between a concrete fact type and its set of alpha roots.